### PR TITLE
Issue #2816059 - Cart issues after Guest checkout

### PR DESCRIPTION
**Pull request by steveoliver**
See https://www.drupal.org/node/2816059

### CODE REVIEW

File: modules/cart/src/CartSession.php [hunk 1]
```diff
@@ -63,6 +71,10 @@ public function deleteCartId($cart_id) {
       // Remove the empty list to allow the system to clean up empty sessions.
       $this->session->remove('commerce_cart_orders');
     }
+    // Keep a list of completed order for access checks to completed orders.
+    $completed_ids = $this->session->get('commerce_cart_completed_orders', []);
+    $completed_ids = array_merge($completed_ids, [$cart_id]);
+    $this->session->set('commerce_cart_completed_orders', $completed_ids);
```
Comment (mglaman) [inline modules/cart/src/CartSession.php:83]: So should this just go in `::finalizeCart` and make sure `::finalizeCart` doesn't invoke delete? Seems weird we have `deleteCartId` but its actually doing a conversions.

Comment (steveoliver) [reply modules/cart/src/CartSession.php:83]: Good call.

Comment (steveoliver) [reply modules/cart/src/CartSession.php:83]: I ended up moving the handling of completed_orders to a method in CartSession (called addCompletedCartId()).

**Comment by mglaman** [comment]
@bojanz can we get your consensus on this? follows 1.x pattern and fixes the false cart issue on anonymous users.

**Comment by bojanz** [comment]
Committed in d7317f7a696bc32ad3d97bac7a1baa2074236162. Updating the parent issue with an explanation of API differences.

**Pull request closed without merging**